### PR TITLE
Avoid rewriting test files when calling cmake

### PR DESCRIPTION
**Pull request by masterleinad**
Previously, we were changing the timestamp of the generated test files every time we called `cmake` resulting in unnecessary rebuilds. `configure_file` only changes the files if they differ.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Please comment that you are writing a temporary intermediate file and calling `configure_file` to avoid updating timestamps which would trigger unnecessary rebuilds on subsequent cmake runs.

**Comment by masterleinad** [comment]
> Please comment that you are writing a temporary intermediate file and calling `configure_file` to avoid updating timestamps which would trigger unnecessary rebuilds on subsequent cmake runs.

See [`8aedb69` (#3548)](https://github.com/kokkos/kokkos/pull/3548/commits/8aedb6953a98fc80073e995f9d66b794270b2019).

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -143,10 +146,13 @@ foreach(Tag Threads;Serial;OpenMP;Cuda;HPX;OpenMPTarget;HIP;SYCL)
...
       )
       set(file ${dir}/Test${Tag}_${Name}.cpp)
-      file(WRITE ${file}
+      # Write to a temporary intermediate file and call configure_file to avoid
+      # updating timestamps triggering unnecessary rebuilds on subsequent cmake runs.
+      file(WRITE ${dir}/dummy.cpp
           "#include <Test${Tag}_Category.hpp>\n"
           "#include <Test${Name}.hpp>\n"
       )
+      configure_file(${dir}/dummy.cpp ${file})
```
Comment (jjwilke) [inline core/unit_test/CMakeLists.txt:155]: Would it be better to have a dummy.cpp.in with @TEST@, etc and skip the first `file(WRITE`?

**Review by jjwilke — CHANGES_REQUESTED** [review]
I can approve without the suggested change - but curious what you think.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Your last commit invalidates the comments.  You must fix it.

I do not like your implementation of the changes suggested by @jjwilke.  One now has to open `test.cpp.in` to understand what is going on and nothing in the name of the variables indicates that they are being substituted in the file being configured.
My preference goes to revert.

**Comment by masterleinad** [comment]
I don't really care one way or the other. @dalg24 seems to feel more strongly than @jjwilke so I'll just revert.

**Comment by masterleinad** [comment]
Retest this please.

**Review by dalg24 — APPROVED** [review]

**Review by jjwilke — APPROVED** [review]
Yep, I'm not going fight on this one : )

**Pull request merged**